### PR TITLE
fix: ⚡️ use nft.storage directory API

### PR DESCRIPTION
**Pull request by melMass**
To get subfolders to work we need to use NFT.storage's API.

The current version flattens subfolders:
image

The PR do not work yet, I'm not sure to understand exactly what is expected by [storeDirectory](https://nftstorage.github.io/nft.storage/client/classes/lib.NFTStorage.html#storeDirectory) from nodejs.

### CODE REVIEW

**Comment by melMass** [comment]
This is what it should look like
image

IPFS: https://nftstorage.link/ipfs/QmPRvZDi9LPVYFgqnG1L2LseHsjuSj5wLon5Zj22zxhus2

**Comment by xat** [comment]
are you sure storeDirectory supports subfolders? i don't see anything about that in the docs.

**Comment by melMass** [comment]
> are you sure storeDirectory supports subfolders? i don't see anything about that in the docs.

Yeah I just realized that too, I asked over their discord, but I'm actually going to check the source code of nft.storage api and adapt, I'm just a bit lost with Buffer/File between nodejs and js

**Comment by xat** [comment]
just found this example here => https://gist.github.com/tougerthao/9452b85fcb1f2b2b129c43dd5ecdc885

**Comment by melMass** [comment]
> just found this example here => [gist.github.com/tougerthao/9452b85fcb1f2b2b129c43dd5ecdc885](https://gist.github.com/tougerthao/9452b85fcb1f2b2b129c43dd5ecdc885)

Yep my actual issue is `getFilesFromPath` from Mutler. If I use plainly req.files (the API doc suggests that should work as it's a FileList) this doesn't work and errors out (file.stream undefined or something like that).
So I have no idea how to build the files object this functions expects

**Comment by xat** [comment]
looks like `storeDirectory` expects this kind of format as input. i suppose the file array would need to be transformed into that.

```
[
    { name: '/path/to/me1', stream: [Function: stream] },
    { name: '/path/to/me2', stream: [Function: stream] }
]
```

**Comment by xat** [comment]
buffer to stream:

```
import { Readable } from 'stream'
const stream = Readable.from(buffer);
```

**Comment by yusefnapora** [comment]
`storeDirectory` uses the path in each file's `name` property to determine the directory structure. It looks like the current code is stripping out the filename by converting to `Blob`...

If you can pull in the [@web-std/file package](https://www.npmjs.com/package/@web-std/file) and convert `req.files` to the `File` objects from that package, it should hopefully work. I think that's what we're using to polyfill `File` in node environments.

Basically instead of this:

```js
const cid = await client.storeDirectory(req.files.map((file) => (
      new Blob(file.buffer)
      )));
```

Do this:

```js
import { File } from "@web-std/file"

const files = req.files.map((file) => new File([file.buffer], file.name))

const cid = await client.storeDirectory(files)
```


I haven't tested the above, so :crossed_fingers: - feel free to ping me if it doesn't work tho.

**Comment by yusefnapora** [comment]
Actually, I just remembered that you should be able to import `File` from the `nft.storage` package, instead of pulling in `@web-std/file`.

So you should be able to do:

```js
import { File, NFTStorage } from 'nft.storage'
```

And get a compatible `File` object.

**Comment by melMass** [comment]
Thanks, this is using the last commit, It's still flattened somehow, I'm going to try to pass the {name,stream} object.

image
